README.md
# circular_countdown

A package that represents a circular countdown that you can use for every kind of units (days, hours, seconds, points, ...).

## Demo

A web demo is available here : https://mattisbrizard.github.io/circular_countdown/#/

[![Build Status](https://img.shields.io/github/workflow/status/MattisBrizard/circular_countdown/Deploy%20Flutter%20web)](https://github.com/MattisBrizard/circular_countdown/actions?query=workflow%3A%22Deploy+Flutter+web%22)

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  circular_countdown: ^2.1.0
```

In your library add the following import:

```dart
import  'package:circular_countdown/circular_countdown.dart';
```

## Static countdown

This is the way to create a simple static countdown.

```dart
import 'package:flutter/material.dart';
...
return CircularCountdown(
	countdownTotal:  10,
	countdownRemaining:  7,
);
...
```

<img width="250"  src="https://raw.githubusercontent.com/MattisBrizard/circular_countdown/master/doc/images/static.png">

By playing with the parameters, you can change the size, the colors, the size of the gaps, the strokeWidth and more to have more complex Countdowns.

<img width="250"  src="https://raw.githubusercontent.com/MattisBrizard/circular_countdown/master/doc/images/complex.png">

## Timer countdown

If you want to have full control on the timer (start, pause, etc), please see the section _Controlled countdown_ below.

If you want to use the `CircularCountdown` widget as a Time Countdown, you can use the built-in `TimeCircularCountdown` that wraps the timer logic for you: (see how in example)

```dart
return TimeCircularCountdown(
    unit: CountdownUnit.seconds,
    countdownTotal: 30,
    onUpdated: (unit, remainingTime) => print('Updated'),
    onFinished: () => print('Countdown finished'),
);
```

<img width="250"  src="https://raw.githubusercontent.com/MattisBrizard/circular_countdown/master/doc/gifs/timer.gif">

You can repeat the timer when ended by using the `repeat` bool property.

## Controlled countdown

To have full control on the timer (play, pause, reset, shared between mutliple screen, etc), you can use `CircularCountdown` with the [timer_controller](https://pub.dev/packages/timer_controller) library.

You can see the code in the [example](https://github.com/MattisBrizard/circular_countdown/tree/master/example/lib/countdowns/controlled_countdown.dart) app.

## Contribution

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue](https://github.com/MattisBrizard/circular_countdown/issues).
If you fixed a bug or implemented a feature, please send a pull [request](https://github.com/MattisBrizard/circular_countdown/pulls).


CHANGELOG.md
# [2.1.0] - 05/05/2021

## Changes :

- `CircularCountdown` is no more wrapped with `Center`.
- Updated README and added an example with [timer_controller](https://pub.dev/packages/timer_controller).

# [2.0.0] - 20/04/2021

## Changes :

- Added null-safety support. (https://github.com/MattisBrizard/circular_countdown/issues/5)

# [1.2.1] - 16/07/2020

## Changes :

- Fixed `TimeCircularCountdown` to care about widget configuration changes for `unit` and `countdownTotal` properties. (https://github.com/MattisBrizard/circular_countdown/issues/3)

# [1.2.0] - 15/07/2020

## Changes :

- Added a `repeat` bool property to `TimeCircularCountdown` to restart countdown when ended (Default to `false`).

```dart
TimeCircularCountdown(
    unit: CountdownUnit.seconds,
    countdownTotal: 5,
    repeat: true,
);
```

# [1.1.1] - 08/07/2020

## Changes :

- Fixed an issue where isClockwise wasn't correctly used.

# [1.1.0] - 08/07/2020

## Changes :

- Added a isClockwise property to draw the countdown clockwise or not.

# [1.0.2] - 28/06/2020

## Changes :

- Added timer gif in Readme.

# [1.0.1] - 14/05/2020

## Changes :

- Updated Documentation.

# [1.0.0] - 10/05/2020

## Changes :

- Added a `TimeCircularCountdown` which provides an easy way to setup a time countdown with the `CircularCountdown`.

- Updated `CircularCountdown` constructor :

  - `diameter` is no longer required. (If not provided then should take either the maximum space if possible or 100)
  - `countdownRemaining` is no longer required.

- Updated README.md

```dart
TimeCircularCountdown(
    unit: CountdownUnit.minute,
    countdownTotal: 5,
    onUpdated: (unit, remainingTime) => print('Updated'),
    onFinished: () => print('Countdown finished'),
    onCanceled: (unit, remainingTime) => print('Canceled'),
    textStyle: const TextStyle(
        color: Colors.white,
        fontSize: 90,
    ),
);
```

## Breaking changes :

- Updated `CircularCountdown` constructor :

  - `textSpan` has been replaced by `textSyle` that will be used to display the current value of the countdown.

# [0.1.0] - 15/08/2019

## Changes :

- Added possibility to display a textSpan in the center of the widget.
  This textSpan will displays only if there is enough space.

```dart
CircularCountdown(
    diameter: 250,
    countdownTotal: 10,
    countdownRemaining: 7,
    strokeWidth: 45,
    textSpan: TextSpan(
        text: '7',
        style: TextStyle(
            color: Colors.white,
            fontSize: 90,
        ),
    ),
)
```

# [0.0.2] - 02/08/2019

Updated doc.

# [0.0.1] - 02/08/2019

Initial release.


example/lib/countdowns/time_countdown.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:circular_countdown/circular_countdown.dart';

class TimeCountdownPage extends StatefulWidget {
  const TimeCountdownPage({
    Key? key,
  }) : super(key: key);
  @override
  _TimeCountdownPageState createState() => _TimeCountdownPageState();
}

class _TimeCountdownPageState extends State<TimeCountdownPage> {
  String description = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Time CountdownPage'),
        backgroundColor: Colors.transparent,
        elevation: 0,
      ),
      backgroundColor: Colors.grey[900],
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            TimeCircularCountdown(
              unit: CountdownUnit.second,
              countdownTotal: 15,
              diameter: 300,
              countdownCurrentColor: Colors.amber,
              onUpdated: (unit, remainingTime) {
                setState(() {
                  description = 'Updated: ${_formatTime(unit, remainingTime)}';
                });
              },
              onFinished: () {
                description = 'Finished';
                setState(() {});
              },
              onCanceled: (unit, remainingTime) => print(
                'Canceled at ${_formatTime(unit, remainingTime)}',
              ),
              textStyle: const TextStyle(
                color: Colors.white,
                fontSize: 90,
              ),
            ),
            Text(
              description,
              style: const TextStyle(
                color: Colors.white,
                fontSize: 30,
              ),
            ),
          ],
        ),
      ),
    );
  }

  String _formatTime(CountdownUnit unit, int remainingTime) =>
      '$remainingTime ${describeEnum(unit)}${remainingTime > 1 ? 's' : ''}';
}
